File: spike/connection_race.py
"""Pure connection-race orchestration with no DNS, clock, or socket I/O."""

from dataclasses import dataclass, replace
from typing import Optional

import address_attempts
import route_circuit


PHASE_RESOLVING = "resolving"
PHASE_CONNECTING = "connecting"
PHASE_CONNECTED = "connected"
PHASE_REJECTED = "rejected"
PHASE_FAILED = "failed"
PHASE_TIMED_OUT = "timed_out"
PHASE_EXHAUSTED = "exhausted"

EVENT_RESOLVED = "resolved"
EVENT_RESOLVE_FAILED = "resolve_failed"
EVENT_ATTEMPT_SUCCEEDED = "attempt_succeeded"
EVENT_ATTEMPT_FAILED = "attempt_failed"
EVENT_WAKE = "wake"

COMMAND_RESOLVE = "resolve"
COMMAND_START = "start"
COMMAND_CANCEL = "cancel"
COMMAND_WAKE = "wake"

TERMINAL_PHASES = frozenset(
    (
        PHASE_CONNECTED,
        PHASE_REJECTED,
        PHASE_FAILED,
        PHASE_TIMED_OUT,
        PHASE_EXHAUSTED,
    )
)


@dataclass(frozen=True)
class ConnectionRaceConfig:
    timeout_ms: int
    stagger_ms: int
    max_concurrent: int
    preferred_family: str


@dataclass(frozen=True)
class ConnectionRaceState:
    key: route_circuit.RouteCircuitKey
    phase: str
    started_at_ms: int
    updated_at_ms: int
    deadline_at_ms: int
    candidates: tuple = ()
    attempts: tuple = ()
    winner_candidate_id: str = ""
    reason: str = ""


@dataclass(frozen=True)
class ConnectionRaceEvent:
    kind: str
    now_ms: int
    candidate_id: str = ""
    candidates: tuple = ()


@dataclass(frozen=True)
class ConnectionRaceCommand:
    kind: str
    candidate_id: str = ""
    at_ms: Optional[int] = None


@dataclass(frozen=True)
class ConnectionRaceTransition:
    state: ConnectionRaceState
    circuit_states: dict
    commands: tuple = ()
    circuit_decisions: tuple = ()


def _validate_config(config):
    if config.timeout_ms < 1:
        raise ValueError("timeout_ms must be positive")
    if config.stagger_ms < 0:
        raise ValueError("stagger_ms must not be negative")
    if config.max_concurrent < 1:
        raise ValueError("max_concurrent must be positive")
    if config.preferred_family not in (
        address_attempts.FAMILY_IPV4,
        address_attempts.FAMILY_IPV6,
    ):
        raise ValueError("preferred_family must be ipv4 or ipv6")


def _address_context(state, config, now_ms):
    return address_attempts.AddressPlanContext(
        now_ms=now_ms,
        started_at_ms=state.started_at_ms,
        deadline_at_ms=state.deadline_at_ms,
        stagger_ms=config.stagger_ms,
        max_concurrent=config.max_concurrent,
        preferred_family=config.preferred_family,
    )


def _record_terminal(
    circuit_states,
    state,
    circuit_config,
    now_ms,
    phase,
    reason,
    winner_candidate_id="",
):
    event_kind = (
        route_circuit.EVENT_RECORD_SUCCESS
        if phase == PHASE_CONNECTED
        else route_circuit.EVENT_RECORD_FAILURE
    )
    updated_circuits, decision = route_circuit.reduce_route_circuit(
        circuit_states,
        route_circuit.CircuitEvent(event_kind, state.key, now_ms),
        circuit_config,
    )
    terminal = replace(
        state,
        phase=phase,
        updated_at_ms=now_ms,
        winner_candidate_id=winner_candidate_id,
        reason=reason,
    )
    return terminal, updated_circuits, decision


def _cancel_attempts(attempts, candidate_ids, now_ms):
    cancelled = set(candidate_ids)
    return tuple(
        replace(
            attempt,
            state=address_attempts.ATTEMPT_CANCELLED,
            completed_at_ms=now_ms,
        )
        if attempt.candidate_id in cancelled
        and attempt.state == address_attempts.ATTEMPT_RUNNING
        else attempt
        for attempt in attempts
    )


def _settle(circuit_states, state, config, circuit_config, now_ms):
    commands = []
    while True:
        result = address_attempts.plan_address_attempts(
            state.candidates,
            state.attempts,
            _address_context(state, config, now_ms),
        )
        decision = result.decision

        if decision.kind == address_attempts.DECISION_START:
            state = replace(
                state,
                updated_at_ms=now_ms,
                attempts=state.attempts
                + (
                    address_attempts.AddressAttempt(
                        candidate_id=decision.candidate_id,
                        state=address_attempts.ATTEMPT_RUNNING,
                        started_at_ms=now_ms,
                    ),
                ),
            )
            commands.append(
                ConnectionRaceCommand(COMMAND_START, decision.candidate_id)
            )
            continue

        if decision.kind == address_attempts.DECISION_WAIT:
            commands.append(
                ConnectionRaceCommand(COMMAND_WAKE, at_ms=decision.wake_at_ms)
            )
            return ConnectionRaceTransition(state, dict(circuit_states), tuple(commands))

        if decision.kind == address_attempts.DECISION_SELECT:
            cancel_commands = tuple(
                ConnectionRaceCommand(COMMAND_CANCEL, candidate_id)
                for candidate_id in decision.cancel
            )
            state = replace(
                state,
                attempts=_cancel_attempts(state.attempts, decision.cancel, now_ms),
            )
            state, circuit_states, circuit_decision = _record_terminal(
                circuit_states,
                state,
                circuit_config,
                now_ms,
                PHASE_CONNECTED,
                "connected",
                decision.candidate_id,
            )
            return ConnectionRaceTransition(
                state,
                circuit_states,
                tuple(commands) + cancel_commands,
                (circuit_decision,),
            )

        if decision.kind == address_attempts.DECISION_TIMEOUT:
            cancel_commands = tuple(
                ConnectionRaceCommand(COMMAND_CANCEL, candidate_id)
                for candidate_id in decision.cancel
            )
            state = replace(
                state,
                attempts=_cancel_attempts(state.attempts, decision.cancel, now_ms),
            )
            state, circuit_states, circuit_decision = _record_terminal(
                circuit_states,
                state,
                circuit_config,
                now_ms,
                PHASE_TIMED_OUT,
                "deadline",
            )
            return ConnectionRaceTransition(
                state,
                circuit_states,
                tuple(commands) + cancel_commands,
                (circuit_decision,),
            )

        if decision.kind == address_attempts.DECISION_EXHAUSTED:
            state, circuit_states, circuit_decision = _record_terminal(
                circuit_states,
                state,
                circuit_config,
                now_ms,
                PHASE_EXHAUSTED,
                "all_attempts_failed",
            )
            return ConnectionRaceTransition(
                state,
                circuit_states,
                tuple(commands),
                (circuit_decision,),
            )

        raise ValueError("unknown address-plan decision")


def start_connection_race(circuit_states, key, config, circuit_config, now_ms):
    """Gate one logical request before asking an adapter to resolve it."""
    _validate_config(config)
    deadline_at_ms = now_ms + config.timeout_ms
    state = ConnectionRaceState(
        key=key,
        phase=PHASE_RESOLVING,
        started_at_ms=now_ms,
        updated_at_ms=now_ms,
        deadline_at_ms=deadline_at_ms,
    )
    updated_circuits, decision = route_circuit.reduce_route_circuit(
        circuit_states,
        route_circuit.CircuitEvent(
            route_circuit.EVENT_BEFORE_REQUEST, key, now_ms
        ),
        circuit_config,
    )
    if decision.kind == route_circuit.DECISION_REJECT:
        return ConnectionRaceTransition(
            replace(state, phase=PHASE_REJECTED, reason=decision.reason),
            updated_circuits,
            circuit_decisions=(decision,),
        )
    return ConnectionRaceTransition(
        state,
        updated_circuits,
        (
            ConnectionRaceCommand(COMMAND_RESOLVE),
            ConnectionRaceCommand(COMMAND_WAKE, at_ms=deadline_at_ms),
        ),
        (decision,),
    )


def _replace_running_attempt(state, candidate_id, attempt_state, now_ms):
    matched = False
    attempts = []
    for attempt in state.attempts:
        if attempt.candidate_id != candidate_id:
            attempts.append(attempt)
            continue
        matched = True
        if attempt.state != address_attempts.ATTEMPT_RUNNING:
            raise ValueError("attempt completion requires a running candidate")
        attempts.append(
            replace(
                attempt,
                state=attempt_state,
                completed_at_ms=now_ms,
            )
        )
    if not matched:
        raise ValueError("attempt completion references an unknown candidate")
    return replace(state, attempts=tuple(attempts), updated_at_ms=now_ms)


def _advance_resolving(circuit_states, state, event, config, circuit_config):
    if event.kind == EVENT_WAKE:
        if event.now_ms < state.deadline_at_ms:
            return ConnectionRaceTransition(
                replace(state, updated_at_ms=event.now_ms),
                dict(circuit_states),
                (ConnectionRaceCommand(COMMAND_WAKE, at_ms=state.deadline_at_ms),),
            )
        state, circuit_states, decision = _record_terminal(
            circuit_states,
            state,
            circuit_config,
            event.now_ms,
            PHASE_TIMED_OUT,
            "resolver_deadline",
        )
        return ConnectionRaceTransition(
            state, circuit_states, circuit_decisions=(decision,)
        )
    if event.kind == EVENT_RESOLVE_FAILED:
        state, circuit_states, decision = _record_terminal(
            circuit_states,
            state,
            circuit_config,
            event.now_ms,
            PHASE_FAILED,
            "resolve_failed",
        )
        return ConnectionRaceTransition(
            state, circuit_states, circuit_decisions=(decision,)
        )
    if event.kind != EVENT_RESOLVED:
        raise ValueError("resolver phase accepts only resolver events or wake")
    state = replace(
        state,
        phase=PHASE_CONNECTING,
        updated_at_ms=event.now_ms,
        candidates=tuple(event.candidates),
    )
    return _settle(circuit_states, state, config, circuit_config, event.now_ms)


def _advance_connecting(circuit_states, state, event, config, circuit_config):
    if event.kind == EVENT_WAKE:
        state = replace(state, updated_at_ms=event.now_ms)
    elif event.kind == EVENT_ATTEMPT_SUCCEEDED:
        state = _replace_running_attempt(
            state,
            event.candidate_id,
            address_attempts.ATTEMPT_SUCCEEDED,
            event.now_ms,
        )
    elif event.kind == EVENT_ATTEMPT_FAILED:
        state = _replace_running_attempt(
            state,
            event.candidate_id,
            address_attempts.ATTEMPT_FAILED,
            event.now_ms,
        )
    else:
        raise ValueError("connecting phase accepts only attempt events or wake")
    return _settle(circuit_states, state, config, circuit_config, event.now_ms)


def advance_connection_race(
    circuit_states,
    state,
    event,
    config,
    circuit_config,
):
    """Apply one adapter event and return deterministic follow-up commands."""
    _validate_config(config)
    if state.phase in TERMINAL_PHASES:
        return ConnectionRaceTransition(state, dict(circuit_states))
    if event.now_ms < state.updated_at_ms:
        raise ValueError("connection-race events must be monotonic")
    if state.phase == PHASE_RESOLVING:
        return _advance_resolving(
            circuit_states, state, event, config, circuit_config
        )
    if state.phase == PHASE_CONNECTING:
        return _advance_connecting(
            circuit_states, state, event, config, circuit_config
        )
    raise ValueError("unknown connection-race phase")

File: app-tauri/src-tauri/src/connection_race.rs
//! Pure connection-race orchestration with no DNS, clock, or socket I/O.

use crate::address_attempts::{
    plan_address_attempts, AddressAttempt, AddressAttemptState, AddressCandidate,
    AddressDecisionKind, AddressFamily, AddressPlanContext,
};
use crate::route_circuit::{
    reduce_route_circuit, CircuitConfig, CircuitDecision, CircuitDecisionKind, CircuitEvent,
    CircuitEventKind, CircuitStates, RouteCircuitKey,
};
use serde::{Deserialize, Serialize};

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct ConnectionRaceConfig {
    pub timeout_ms: u64,
    pub stagger_ms: u64,
    pub max_concurrent: usize,
    pub preferred_family: AddressFamily,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ConnectionRacePhase {
    Resolving,
    Connecting,
    Connected,
    Rejected,
    Failed,
    TimedOut,
    Exhausted,
}

impl ConnectionRacePhase {
    pub fn is_terminal(self) -> bool {
        matches!(
            self,
            Self::Connected | Self::Rejected | Self::Failed | Self::TimedOut | Self::Exhausted
        )
    }
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct ConnectionRaceState {
    pub key: RouteCircuitKey,
    pub phase: ConnectionRacePhase,
    pub started_at_ms: u64,
    pub updated_at_ms: u64,
    pub deadline_at_ms: u64,
    pub candidates: Vec<AddressCandidate>,
    pub attempts: Vec<AddressAttempt>,
    pub winner_candidate_id: String,
    pub reason: String,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ConnectionRaceEventKind {
    Resolved,
    ResolveFailed,
    AttemptSucceeded,
    AttemptFailed,
    Wake,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct ConnectionRaceEvent {
    pub kind: ConnectionRaceEventKind,
    pub now_ms: u64,
    pub candidate_id: String,
    pub candidates: Vec<AddressCandidate>,
}

impl ConnectionRaceEvent {
    pub fn wake(now_ms: u64) -> Self {
        Self {
            kind: ConnectionRaceEventKind::Wake,
            now_ms,
            candidate_id: String::new(),
            candidates: Vec::new(),
        }
    }
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ConnectionRaceCommandKind {
    Resolve,
    Start,
    Cancel,
    Wake,
}

#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
pub struct ConnectionRaceCommand {
    pub kind: ConnectionRaceCommandKind,
    pub candidate_id: String,
    pub at_ms: Option<u64>,
}

#[derive(Clone, Debug, Eq, PartialEq)]
pub struct ConnectionRaceTransition {
    pub state: ConnectionRaceState,
    pub circuit_states: CircuitStates,
    pub commands: Vec<ConnectionRaceCommand>,
    pub circuit_decisions: Vec<CircuitDecision>,
}

fn command(
    kind: ConnectionRaceCommandKind,
    candidate_id: impl Into<String>,
    at_ms: Option<u64>,
) -> ConnectionRaceCommand {
    ConnectionRaceCommand {
        kind,
        candidate_id: candidate_id.into(),
        at_ms,
    }
}

fn validate_config(config: &ConnectionRaceConfig) -> Result<(), String> {
    if config.timeout_ms < 1 {
        return Err("timeout_ms must be positive".into());
    }
    if config.max_concurrent < 1 {
        return Err("max_concurrent must be positive".into());
    }
    Ok(())
}

fn address_context(
    state: &ConnectionRaceState,
    config: &ConnectionRaceConfig,
    now_ms: u64,
) -> AddressPlanContext {
    AddressPlanContext {
        now_ms,
        started_at_ms: state.started_at_ms,
        deadline_at_ms: state.deadline_at_ms,
        stagger_ms: config.stagger_ms,
        max_concurrent: config.max_concurrent,
        preferred_family: config.preferred_family,
    }
}

fn record_terminal(
    circuit_states: &CircuitStates,
    mut state: ConnectionRaceState,
    circuit_config: &CircuitConfig,
    now_ms: u64,
    phase: ConnectionRacePhase,
    reason: &str,
    winner_candidate_id: &str,
) -> Result<(ConnectionRaceState, CircuitStates, CircuitDecision), String> {
    let kind = if phase == ConnectionRacePhase::Connected {
        CircuitEventKind::RecordSuccess
    } else {
        CircuitEventKind::RecordFailure
    };
    let (updated_circuits, decision) = reduce_route_circuit(
        circuit_states,
        &CircuitEvent {
            kind,
            key: state.key.clone(),
            now_ms,
        },
        circuit_config,
    )?;
    state.phase = phase;
    state.updated_at_ms = now_ms;
    state.winner_candidate_id = winner_candidate_id.to_string();
    state.reason = reason.to_string();
    Ok((state, updated_circuits, decision))
}

fn cancel_attempts(state: &mut ConnectionRaceState, candidate_ids: &[String], now_ms: u64) {
    for attempt in &mut state.attempts {
        if candidate_ids.contains(&attempt.candidate_id)
            && attempt.state == AddressAttemptState::Running
        {
            attempt.state = AddressAttemptState::Cancelled;
            attempt.completed_at_ms = Some(now_ms);
        }
    }
}

fn settle(
    circuit_states: &CircuitStates,
    mut state: ConnectionRaceState,
    config: &ConnectionRaceConfig,
    circuit_config: &CircuitConfig,
    now_ms: u64,
) -> Result<ConnectionRaceTransition, String> {
    let mut commands = Vec::new();
    loop {
        let result = plan_address_attempts(
            &state.candidates,
            &state.attempts,
            &address_context(&state, config, now_ms),
        )?;
        match result.decision.kind {
            AddressDecisionKind::Start => {
                let candidate_id = result.decision.candidate_id;
                state.updated_at_ms = now_ms;
                state.attempts.push(AddressAttempt {
                    candidate_id: candidate_id.clone(),
                    state: AddressAttemptState::Running,
                    started_at_ms: now_ms,
                    completed_at_ms: None,
                });
                commands.push(command(
                    ConnectionRaceCommandKind::Start,
                    candidate_id,
                    None,
                ));
            }
            AddressDecisionKind::Wait => {
                commands.push(command(
                    ConnectionRaceCommandKind::Wake,
                    "",
                    result.decision.wake_at_ms,
                ));
                return Ok(ConnectionRaceTransition {
                    state,
                    circuit_states: circuit_states.clone(),
                    commands,
                    circuit_decisions: Vec::new(),
                });
            }
            AddressDecisionKind::Select => {
                cancel_attempts(&mut state, &result.decision.cancel, now_ms);
                commands.extend(result.decision.cancel.iter().map(|candidate_id| {
                    command(
                        ConnectionRaceCommandKind::Cancel,
                        candidate_id.clone(),
                        None,
                    )
                }));
                let (state, circuit_states, decision) = record_terminal(
                    circuit_states,
                    state,
                    circuit_config,
                    now_ms,
                    ConnectionRacePhase::Connected,
                    "connected",
                    &result.decision.candidate_id,
                )?;
                return Ok(ConnectionRaceTransition {
                    state,
                    circuit_states,
                    commands,
                    circuit_decisions: vec![decision],
                });
            }
            AddressDecisionKind::Timeout => {
                cancel_attempts(&mut state, &result.decision.cancel, now_ms);
                commands.extend(result.decision.cancel.iter().map(|candidate_id| {
                    command(
                        ConnectionRaceCommandKind::Cancel,
                        candidate_id.clone(),
                        None,
                    )
                }));
                let (state, circuit_states, decision) = record_terminal(
                    circuit_states,
                    state,
                    circuit_config,
                    now_ms,
                    ConnectionRacePhase::TimedOut,
                    "deadline",
                    "",
                )?;
                return Ok(ConnectionRaceTransition {
                    state,
                    circuit_states,
                    commands,
                    circuit_decisions: vec![decision],
                });
            }
            AddressDecisionKind::Exhausted => {
                let (state, circuit_states, decision) = record_terminal(
                    circuit_states,
                    state,
                    circuit_config,
                    now_ms,
                    ConnectionRacePhase::Exhausted,
                    "all_attempts_failed",
                    "",
                )?;
                return Ok(ConnectionRaceTransition {
                    state,
                    circuit_states,
                    commands,
                    circuit_decisions: vec![decision],
                });
            }
        }
    }
}

pub fn start_connection_race(
    circuit_states: &CircuitStates,
    key: RouteCircuitKey,
    config: &ConnectionRaceConfig,
    circuit_config: &CircuitConfig,
    now_ms: u64,
) -> Result<ConnectionRaceTransition, String> {
    validate_config(config)?;
    let deadline_at_ms = now_ms
        .checked_add(config.timeout_ms)
        .ok_or_else(|| "connection-race deadline overflow".to_string())?;
    let mut state = ConnectionRaceState {
        key: key.clone(),
        phase: ConnectionRacePhase::Resolving,
        started_at_ms: now_ms,
        updated_at_ms: now_ms,
        deadline_at_ms,
        candidates: Vec::new(),
        attempts: Vec::new(),
        winner_candidate_id: String::new(),
        reason: String::new(),
    };
    let (updated_circuits, decision) = reduce_route_circuit(
        circuit_states,
        &CircuitEvent {
            kind: CircuitEventKind::BeforeRequest,
            key,
            now_ms,
        },
        circuit_config,
    )?;
    if decision.kind == CircuitDecisionKind::Reject {
        state.phase = ConnectionRacePhase::Rejected;
        state.reason = decision.reason.clone();
        return Ok(ConnectionRaceTransition {
            state,
            circuit_states: updated_circuits,
            commands: Vec::new(),
            circuit_decisions: vec![decision],
        });
    }
    Ok(ConnectionRaceTransition {
        state,
        circuit_states: updated_circuits,
        commands: vec![
            command(ConnectionRaceCommandKind::Resolve, "", None),
            command(ConnectionRaceCommandKind::Wake, "", Some(deadline_at_ms)),
        ],
        circuit_decisions: vec![decision],
    })
}

fn replace_running_attempt(
    mut state: ConnectionRaceState,
    candidate_id: &str,
    attempt_state: AddressAttemptState,
    now_ms: u64,
) -> Result<ConnectionRaceState, String> {
    let attempt = state
        .attempts
        .iter_mut()
        .find(|attempt| attempt.candidate_id == candidate_id)
        .ok_or_else(|| "attempt completion references an unknown candidate".to_string())?;
    if attempt.state != AddressAttemptState::Running {
        return Err("attempt completion requires a running candidate".into());
    }
    attempt.state = attempt_state;
    attempt.completed_at_ms = Some(now_ms);
    state.updated_at_ms = now_ms;
    Ok(state)
}

fn defer_deadline_tie(
    circuit_states: &CircuitStates,
    mut state: ConnectionRaceState,
    event: &ConnectionRaceEvent,
) -> Option<ConnectionRaceTransition> {
    let has_running_attempt = state
        .attempts
        .iter()
        .any(|attempt| attempt.state == AddressAttemptState::Running);
    if event.kind != ConnectionRaceEventKind::Wake
        || event.now_ms != state.deadline_at_ms
        || !has_running_attempt
    {
        return None;
    }
    let wake_at_ms = event.now_ms.checked_add(1)?;
    state.updated_at_ms = event.now_ms;
    Some(ConnectionRaceTransition {
        state,
        circuit_states: circuit_states.clone(),
        commands: vec![command(
            ConnectionRaceCommandKind::Wake,
            "",
            Some(wake_at_ms),
        )],
        circuit_decisions: Vec::new(),
    })
}

fn advance_resolving(
    circuit_states: &CircuitStates,
    mut state: ConnectionRaceState,
    event: &ConnectionRaceEvent,
    config: &ConnectionRaceConfig,
    circuit_config: &CircuitConfig,
) -> Result<ConnectionRaceTransition, String> {
    match event.kind {
        ConnectionRaceEventKind::Wake if event.now_ms < state.deadline_at_ms => {
            state.updated_at_ms = event.now_ms;
            let deadline_at_ms = state.deadline_at_ms;
            Ok(ConnectionRaceTransition {
                state,
                circuit_states: circuit_states.clone(),
                commands: vec![command(
                    ConnectionRaceCommandKind::Wake,
                    "",
                    Some(deadline_at_ms),
                )],
                circuit_decisions: Vec::new(),
            })
        }
        ConnectionRaceEventKind::Wake => {
            let (state, circuit_states, decision) = record_terminal(
                circuit_states,
                state,
                circuit_config,
                event.now_ms,
                ConnectionRacePhase::TimedOut,
                "resolver_deadline",
                "",
            )?;
            Ok(ConnectionRaceTransition {
                state,
                circuit_states,
                commands: Vec::new(),
                circuit_decisions: vec![decision],
            })
        }
        ConnectionRaceEventKind::ResolveFailed => {
            let (state, circuit_states, decision) = record_terminal(
                circuit_states,
                state,
                circuit_config,
                event.now_ms,
                ConnectionRacePhase::Failed,
                "resolve_failed",
                "",
            )?;
            Ok(ConnectionRaceTransition {
                state,
                circuit_states,
                commands: Vec::new(),
                circuit_decisions: vec![decision],
            })
        }
        ConnectionRaceEventKind::Resolved => {
            state.phase = ConnectionRacePhase::Connecting;
            state.updated_at_ms = event.now_ms;
            state.candidates = event.candidates.clone();
            settle(circuit_states, state, config, circuit_config, event.now_ms)
        }
        _ => Err("resolver phase accepts only resolver events or wake".into()),
    }
}

fn advance_connecting(
    circuit_states: &CircuitStates,
    state: ConnectionRaceState,
    event: &ConnectionRaceEvent,
    config: &ConnectionRaceConfig,
    circuit_config: &CircuitConfig,
) -> Result<ConnectionRaceTransition, String> {
    if let Some(transition) = defer_deadline_tie(circuit_states, state.clone(), event) {
        return Ok(transition);
    }
    if event.kind == ConnectionRaceEventKind::AttemptSucceeded
        && event.now_ms > state.deadline_at_ms
    {
        return settle(circuit_states, state, config, circuit_config, event.now_ms);
    }
    let state = match event.kind {
        ConnectionRaceEventKind::Wake => {
            let mut state = state;
            state.updated_at_ms = event.now_ms;
            state
        }
        ConnectionRaceEventKind::AttemptSucceeded => replace_running_attempt(
            state,
            &event.candidate_id,
            AddressAttemptState::Succeeded,
            event.now_ms,
        )?,
        ConnectionRaceEventKind::AttemptFailed => replace_running_attempt(
            state,
            &event.candidate_id,
            AddressAttemptState::Failed,
            event.now_ms,
        )?,
        _ => return Err("connecting phase accepts only attempt events or wake".into()),
    };
    settle(circuit_states, state, config, circuit_config, event.now_ms)
}

pub fn advance_connection_race(
    circuit_states: &CircuitStates,
    state: ConnectionRaceState,
    event: &ConnectionRaceEvent,
    config: &ConnectionRaceConfig,
    circuit_config: &CircuitConfig,
) -> Result<ConnectionRaceTransition, String> {
    validate_config(config)?;
    if state.phase.is_terminal() {
        return Ok(ConnectionRaceTransition {
            state,
            circuit_states: circuit_states.clone(),
            commands: Vec::new(),
            circuit_decisions: Vec::new(),
        });
    }
    if event.now_ms < state.updated_at_ms {
        return Err("connection-race events must be monotonic".into());
    }
    match state.phase {
        ConnectionRacePhase::Resolving => {
            advance_resolving(circuit_states, state, event, config, circuit_config)
        }
        ConnectionRacePhase::Connecting => {
            advance_connecting(circuit_states, state, event, config, circuit_config)
        }
        _ => Err("unknown connection-race phase".into()),
    }
}
